### PR TITLE
CDAP-4648 Integrated CustomAction class in the WorkflowDriver.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-4648
This PR adds ability to call the `CustomAction` from the `WorkflowDriver`. 
Pending: Removal of usages of `WorkflowAction` from the code base. will be done as a separate PR.

### CODE REVIEW

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT4354-1

File: cdap-api/src/main/java/co/cask/cdap/api/customaction/CustomActionContext.java [hunk 1]
```diff
@@ -25,7 +25,7 @@
 /**
  * Represents runtime context of the {@link CustomAction} in the Workflow.
  */
-public interface CustomActionContext extends RuntimeContext, Transactional, PluginContext, WorkflowInfoProvider {
+public interface CustomActionContext extends RuntimeContext, Transactional, WorkflowInfoProvider {
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/customaction/CustomActionContext.java:28]: why is PluginContext getting removed? Without it, we cannot use it in apps like Hydrator.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/customaction/BasicCustomActionContext.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    return state;
+  }
+
+  @Override
+  public void execute(TxRunnable runnable) throws TransactionFailureException {
+    final TransactionContext context = datasetCache.newTransactionContext();
+    try {
+      context.start();
+      runnable.run(datasetCache);
+      context.finish();
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/customaction/BasicCustomActionContext.java:127]: does this need to be run in a finally? or is it handled by the abortTransaction calls?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java [hunk 3]
```diff
@@ -75,8 +104,28 @@ private WorkflowAction createAction(BasicWorkflowContext context, InstantiatorFa
     return action;
   }
 
+  @SuppressWarnings("unchecked")
+  private CustomAction createCustomAction(BasicCustomActionContext context, InstantiatorFactory instantiator,
+                                            ClassLoader classLoader) throws Exception {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java]: nit: indentation

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramInfo.java [hunk 4]
```diff
@@ -61,7 +61,7 @@ public static WorkflowProgramInfo create(Arguments arguments) {
                                    programNameInWorkflow, GSON.fromJson(workflowToken, BasicWorkflowToken.class));
   }
 
-  private WorkflowProgramInfo(String workflowName, String workflowNodeId,
+  WorkflowProgramInfo(String workflowName, String workflowNodeId,
                               String workflowRunId, String programNameInWorkflow, BasicWorkflowToken workflowToken) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramInfo.java]: nit: indentation

**Comment by albertshau** [comment]
is the implementation of PluginContext going to be added back in a separate PR? If so, lgtm after small nitpicks.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/customaction/BasicCustomActionContext.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/customaction/BasicCustomActionContext.java:127]: `context.finish()` commits the transaction, which only need to happen if the `run` method succeeds. If there is failure in the execution of the run method, we abort the transaction using `context.abort()`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java]: fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramInfo.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowProgramInfo.java]: fixed.

**Comment by sagarkapare** [comment]
@albertshau Addressed comments. Please take a look again. I will do the implementation of the `PluginContext` in a separate PR where I will make changes to `DataPipelineApp` to include the custom actions.

**Comment by sagarkapare** [comment]
Build is green - http://builds.cask.co/browse/CDAP-DUT4354-1

**Comment by albertshau** [comment]
lgtm

**Comment by sagarkapare** [comment]
Thanks @albertshau for the review. Squashing and merging.

**Pull request merged**